### PR TITLE
Revamp index mechanism with a WAL

### PR DESCRIPTION
**Pull request by equals215**
Here's the gist:

- In-memory index: We'll keep the entire index+metadata in memory for fast read/write operations.
- Write-Ahead Log (WAL): Every index change is immediately logged to a WAL on disk. This ensures durability of recent changes.
- Periodic disk dumps: Every minute, we'll write the entire in-memory index to a separate index file on disk and truncate the WAL.

The workflow looks like this:

- For writes: Update the in-memory index and append to the WAL.
- For reads: Just query the in-memory index.
- Every _n_ time: Dump the in-memory index to disk and reset the WAL.

Key benefits:

- High performance: All operations work with the in-memory index.
- Strong durability: All changes are immediately recorded in the WAL.
- Bounded resource usage: The WAL is truncated regularly.
- Simple recovery: On startup after a crash, load the index file, then replay the WAL, check consistency of the queue with the index.

Consistency check after crash: 

- If an item is in the queue but not in index : Try to recover item then if recovered insert it in the index, else void the item
- If an item is not in the queue but in the index : Delete it from the index

### CODE REVIEW

**Comment by equals215** [comment]
first MVP :
- in-memory index works as expected
- in-memory index successfully dumps into the index file
- WAL is being written as expected

TODO : 
- check that index can recover
- test on long-running process
- add memory usage watching
- **TEST THE F OUT OF INDEX PACKAGE**

**Comment by equals215** [comment]
@CorentinB gob fails to decode metadata at start when a job exists, can you take a look?
To reproduce, just start a crawl with enough seeds so the queue fills well, CTRL+C it once so it exits gracefully, then restart

Maybe and I say maybe we could add stats to the index, but without as much resiliency as the index has
^ this might make no sense tho, brain burnt out tday

**Comment by CorentinB** [comment]
@equals215 didn't we decide to move out of Gob for Protobuf for metadata too?

**Comment by equals215** [comment]
> @equals215 didn't we decide to move out of Gob for Protobuf for metadata too?

argh, you're right
be we should address this in another PR -> this one is only for index
so if we decide to merge metadata into index I will address this problem here

File: internal/pkg/queue/index/file_io.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	// Truncate the WAL file
+	if err := im.walFile.Truncate(0); err != nil {
+		return fmt.Errorf("failed to truncate WAL file: %w", err)
+	}
+	if _, err := im.walFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("failed to seek WAL file: %w", err)
+	}
+
+	im.opsSinceDump = 0
+	im.lastDumpTime = time.Now()
```
[the remark this answers is not in the document]
Comment (equals215) [reply internal/pkg/queue/index/file_io.go:68]: we never use UTC AFAIK

File: internal/pkg/queue/index/file_io.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	defer tempFile.Close()
+
+	// Dump the current index to the temporary file
+	if err := im.dumpIndexToFile(tempFile); err != nil {
+		return fmt.Errorf("failed to dump index to temp file: %w", err)
+	}
+
+	// Remove any previous backup index file
+	if _, err := os.Stat(im.indexFile.Name() + ".old"); err == nil {
```
[the remark this answers is not in the document]
Comment (equals215) [reply internal/pkg/queue/index/file_io.go:41]: To check the file exists before removing, can be discarded

File: internal/pkg/queue/dequeue.go [hunk 3]
```diff
@@ -11,53 +13,46 @@ func (q *PersistentGroupedQueue) Dequeue() (*Item, error) {
...
-		if err != nil {
-			return nil, fmt.Errorf("failed to read item at position %d: %w", positions[0], err)
-		}
+	if len(hosts) == 0 {
+		return nil, ErrNoHostsInQueue
+	}
 
-		item, err := decodeProtoItem(itemBytes)
+	for _, host := range hosts {
+		_, position, size, err = q.index.Pop(host)
```
[the remark this answers is not in the document]
Comment (equals215) [reply internal/pkg/queue/dequeue.go:32]: I don't agree with you on that :
- some usages could use the ID
- you could write benchmarks functions that add item with a given ID and monitor when an item with that ID is popped from queue to get an average queue/dequeue time
- Pop() returns blob{} fields and acts like an interface

Let it keep it's return values and discard when not needed, like many functions do :)

**Pull request merged**